api/internal/handler/data_loader/route_import.go

```diff
@@ -19,535 +19,317 @@ package data_loader
 	}
 
-	// merge route
-	idRoute := make(map[string]*entity.Route)
-	for _, route := range routes {
-		if existRoute, ok := idRoute[route.ID.(string)]; ok {
-			uris := append(existRoute.Uris, route.Uris...)
-			existRoute.Uris = uris
-		} else {
-			idRoute[route.ID.(string)] = route
-		}
-	}
-
-	routes = make([]*entity.Route, 0, len(idRoute))
-	for _, route := range idRoute {
-		routes = append(routes, route)
-	}
-
-	// create route
-	for _, route := range routes {
-		if Force && route.ID != nil {
-			if _, err := h.routeStore.Update(c.Context(), route, true); err != nil {
-				return handler.SpecCodeResponse(err),
-					fmt.Errorf("update route(uris:%v) failed: %s", route.Uris, err)
-			}
-		} else {
-			if _, err := h.routeStore.Create(c.Context(), route); err != nil {
-				return handler.SpecCodeResponse(err),
-					fmt.Errorf("create route(uris:%v) failed: %s", route.Uris, err)
-			}
-		}
-	}
-
-	return map[string]int{
-		"paths":  len(swagger.Paths),
-		"routes": len(routes),
-	}, nil
+	// Create APISIX resources
+	createErrs := h.createEntities(c.Context(), dataSets)
+	return h.convertToImportResult(dataSets, createErrs), nil
 }
 
-func checkRouteExist(ctx context.Context, routeStore *store.GenericStore, route *entity.Route) error {
-	//routeStore := store.GetStore(store.HubKeyRoute)
-	ret, err := routeStore.List(ctx, store.ListInput{
-		Predicate: func(obj interface{}) bool {
-			id := utils.InterfaceToString(route.ID)
-			item := obj.(*entity.Route)
-			if id != "" && id != utils.InterfaceToString(item.ID) {
-				return false
-			}
-
-			itemUris := item.Uris
-			if item.URI != "" {
-				if itemUris == nil {
-					itemUris = []string{item.URI}
-				} else {
-					itemUris = append(itemUris, item.URI)
-				}
-			}
-
-			routeUris := route.Uris
-			if route.URI != "" {
-				if routeUris == nil {
-					routeUris = []string{route.URI}
-				} else {
-					routeUris = append(routeUris, route.URI)
+// Pre-check imported data for duplicates
+// The main problem facing duplication is routing, so here
+// we mainly check the duplication of routes, based on
+// domain name and uri.
+func (h *ImportHandler) preCheck(ctx context.Context, data *loader.DataSets) map[store.HubKey][]string {
+	errs := make(map[store.HubKey][]string)
+	for _, route := range data.Routes {
+		errs[store.HubKeyRoute] = make([]string, 0)
+		o, err := h.routeStore.List(ctx, store.ListInput{
+			// The check logic here is that if when a duplicate HOST or URI
+			// has been found, the HTTP method is checked for overlap, and
+			// if there is overlap it is determined to be a duplicate route
+			// and the import is rejected.
+			Predicate: func(obj interface{}) bool {
+				r := obj.(*entity.Route)
+				isMethodDuplicated := len(intersect.Hash(r.Methods, route.Methods)) > 0
+
+				// Check URI and host duplication
+				// First check for duplicate URIs
+				if (r.URI != "" && route.URI != "" && r.URI == route.URI) || len(intersect.Hash(r.Uris, route.Uris)) > 0 {
```

Comment:
This if condition is complicated. Can we encapsulate a method?

Reply:
yes, let me try

Reply:
They are now drawn out as separate variables to be judged, and the difficulty of reading them has been reduced.